File: experimental/Ornaments.kt
package musics.elements

import musics.common.Ratio
import musics.parts.Leaf

fun ornamented(leaf: Leaf): List<Leaf> {
    val ornamentType = leaf.text.replace("\\", "")
    val leafs = ornamentFunctionMap.getOrDefault(ornamentType, ::plain)(leaf)
    return leafs
}

fun plain(leaf: Leaf): List<Leaf> = listOf(leaf)

fun prall(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    return listOf(
        Leaf(value8, pitches, key, volume, articulation, ties),
        Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties),
        Leaf(duration * Ratio(3, 4), pitches, key, volume, articulation, ties)
    )
}

fun prallup(leaf: Leaf): List<Leaf> = listOf(leaf)
fun pralldown(leaf: Leaf): List<Leaf> = listOf(leaf)

fun upprall(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    val low = Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties)
    return listOf(low, base, high, base, high, base, high, base)
}

fun downprall(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    val low = Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties)
    return listOf(high, base, low, base, high, base, high, base)
}

fun prallprall(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    return listOf(
        base, high, base, high,
        Leaf(duration * Ratio(1, 2), pitches, key, volume, articulation, ties)
    )
}

fun lineprall(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value16 = duration / 16
    val high = Leaf(value16, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value16, pitches, key, volume, articulation, ties)
    return listOf(
        Leaf(duration / 2, pitches.map { localScale.upper(it) }, key, volume, articulation, ties),
        base, high, base, high,
        Leaf(duration / 4, pitches, key, volume, articulation, ties)
    )
}

fun prallmordent(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    val low = Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties)
    return listOf(high, base, high, base, high, base, low, base)
}

fun mordent(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    return listOf(
        Leaf(value8, pitches, key, volume, articulation, ties),
        Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties),
        Leaf(duration * Ratio(3, 4), pitches, key, volume, articulation, ties)
    )
}

fun upmordent(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    val low = Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties)
    return listOf(low, base, high, base, high, base, low, base)
}

fun downmordent(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 12
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    val low = Leaf(value8, pitches.map { localScale.lower(it) }, key, volume, articulation, ties)
    return listOf(
        high, base, low, base,
        high, base, high, base,
        high, base, low, base
    )
}

fun trill(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value8 = duration / 8
    val high = Leaf(value8, pitches.map { localScale.upper(it) }, key, volume, articulation, ties)
    val base = Leaf(value8, pitches, key, volume, articulation, ties)
    return listOf(
        high, base, high, base, high,
        Leaf(duration * Ratio(3, 8), pitches, key, volume, articulation, ties)
    )
}

fun turn(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value4 = duration / 4
    val base = Leaf(value4, pitches, key, volume, articulation, ties)
    return listOf(
        Leaf(value4, pitches.map { localScale.upper(it) }, key, volume, articulation, ties),
        base,
        Leaf(value4, pitches.map { localScale.lower(it) }, key, volume, articulation, ties),
        base
    )
}

fun reverseturn(leaf: Leaf): List<Leaf> = with(leaf) {
    val localScale = key!!.scale
    val value4 = duration / 4
    val base = Leaf(value4, pitches, key, volume, articulation, ties)
    return listOf(
        Leaf(value4, pitches.map { localScale.lower(it) }, key, volume, articulation, ties),
        base,
        Leaf(value4, pitches.map { localScale.upper(it) }, key, volume, articulation, ties),
        base
    )
}

// fermata

fun shortfermata(leaf: Leaf): List<Leaf> = with(leaf) {
    listOf(Leaf(duration * Ratio(3, 2), pitches, key, volume, articulation, ties))
}

fun fermata(leaf: Leaf): List<Leaf> = with(leaf) {
    listOf(Leaf(duration * Ratio(4, 2), pitches, key, volume, articulation, ties))
}

fun longfermata(leaf: Leaf): List<Leaf> = with(leaf) {
    listOf(Leaf(duration * Ratio(6, 2), pitches, key, volume, articulation, ties))
}

fun verylongfermata(leaf: Leaf): List<Leaf> = with(leaf) {
    listOf(Leaf(duration * Ratio(8, 2), pitches, key, volume, articulation, ties))
}

private val ornamentFunctionMap = mapOf<String, (Leaf) -> List<Leaf>>(
    "prall" to ::prall,
    "prallup" to ::prallup,
    "pralldown" to ::pralldown,
    "upprall" to ::upprall,
    "downprall" to ::downprall,
    "prallprall" to ::prallprall,
    "lineprall" to ::lineprall,
    "prallmordent" to ::prallmordent,
    "mordent" to ::mordent,
    "upmordent" to ::upmordent,
    "downmordent" to ::downmordent,
    "trill" to ::trill,
    "turn" to ::turn,
    "reverseturn" to ::reverseturn,
    "shortfermata" to ::shortfermata,
    "fermata" to ::fermata,
    "longfermata" to ::longfermata,
    "verylongfermata" to ::verylongfermata
)

File: src/core/parser/ornaments.py
from __future__ import annotations

from typing import List, Callable, Dict, Any

from core.domain.leafs import Leaf
from tools.ratio import Ratio


# ============================================================
# Ornaments as symbolic transformations
# ============================================================

def ornamented(leaf: Leaf, ornament: str, scale: Any) -> List[Leaf]:
    """
    Apply an ornament to a Leaf using the given scale.

    - leaf:     the base Leaf to ornament
    - ornament: string name (e.g. "\\prall", "trill")
    - scale:    object providing .upper(pitch) and .lower(pitch)
    """
    ornament_type = ornament.replace("\\", "")
    func = ornament_function_map.get(ornament_type, plain)
    return func(leaf, scale)


def plain(leaf: Leaf, scale: Any) -> List[Leaf]:
    """No ornament."""
    return [leaf]


# ------------------------------------------------------------
# Helper: expressive parameter preservation
# ------------------------------------------------------------

def _base_kwargs(leaf: Leaf) -> dict:
    """
    Preserve expressive overrides and context.
    Context-based resolution happens later in the performer.
    """
    return dict(
        dynamic=leaf.dynamic,
        articulation=leaf.articulation,
        timbre=leaf.timbre,
        tied=leaf.tied,
        context=leaf.context,
    )


# ============================================================
# Ornament implementations
# ============================================================

def prall(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    return [
        Leaf(duration=value8, pitches=leaf.pitches, **kw),
        Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw),
        Leaf(duration=leaf.duration * Ratio(3, 4), pitches=leaf.pitches, **kw),
    ]


def prallup(leaf: Leaf, scale: Any) -> List[Leaf]:
    return [leaf]


def pralldown(leaf: Leaf, scale: Any) -> List[Leaf]:
    return [leaf]


def upprall(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    low  = Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw)
    return [low, base, high, base, high, base, high, base]


def downprall(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    low  = Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw)
    return [high, base, low, base, high, base, high, base]


def prallprall(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    return [
        base, high, base, high,
        Leaf(duration=leaf.duration * Ratio(1, 2), pitches=leaf.pitches, **kw),
    ]


def lineprall(leaf: Leaf, scale: Any) -> List[Leaf]:
    value16 = leaf.duration / 16
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value16, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value16, pitches=leaf.pitches, **kw)
    return [
        Leaf(duration=leaf.duration / 2, pitches=[scale.upper(p) for p in leaf.pitches], **kw),
        base, high, base, high,
        Leaf(duration=leaf.duration / 4, pitches=leaf.pitches, **kw),
    ]


def prallmordent(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    low  = Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw)
    return [high, base, high, base, high, base, low, base]


def mordent(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    return [
        Leaf(duration=value8, pitches=leaf.pitches, **kw),
        Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw),
        Leaf(duration=leaf.duration * Ratio(3, 4), pitches=leaf.pitches, **kw),
    ]


def upmordent(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    low  = Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw)
    return [low, base, high, base, high, base, low, base]


def downmordent(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 12
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    low  = Leaf(duration=value8, pitches=[scale.lower(p) for p in leaf.pitches], **kw)
    return [
        high, base, low, base,
        high, base, high, base,
        high, base, low, base,
    ]


def trill(leaf: Leaf, scale: Any) -> List[Leaf]:
    value8 = leaf.duration / 8
    kw = _base_kwargs(leaf)
    high = Leaf(duration=value8, pitches=[scale.upper(p) for p in leaf.pitches], **kw)
    base = Leaf(duration=value8, pitches=leaf.pitches, **kw)
    return [
        high, base, high, base, high,
        Leaf(duration=leaf.duration * Ratio(3, 8), pitches=leaf.pitches, **kw),
    ]


def turn(leaf: Leaf, scale: Any) -> List[Leaf]:
    value4 = leaf.duration / 4
    kw = _base_kwargs(leaf)
    base = Leaf(duration=value4, pitches=leaf.pitches, **kw)
    return [
        Leaf(duration=value4, pitches=[scale.upper(p) for p in leaf.pitches], **kw),
        base,
        Leaf(duration=value4, pitches=[scale.lower(p) for p in leaf.pitches], **kw),
        base,
    ]


def reverseturn(leaf: Leaf, scale: Any) -> List[Leaf]:
    value4 = leaf.duration / 4
    kw = _base_kwargs(leaf)
    base = Leaf(duration=value4, pitches=leaf.pitches, **kw)
    return [
        Leaf(duration=value4, pitches=[scale.lower(p) for p in leaf.pitches], **kw),
        base,
        Leaf(duration=value4, pitches=[scale.upper(p) for p in leaf.pitches], **kw),
        base,
    ]


def shortfermata(leaf: Leaf, scale: Any) -> List[Leaf]:
    kw = _base_kwargs(leaf)
    return [Leaf(duration=leaf.duration * Ratio(3, 2), pitches=leaf.pitches, **kw)]


def fermata(leaf: Leaf, scale: Any) -> List[Leaf]:
    kw = _base_kwargs(leaf)
    return [Leaf(duration=leaf.duration * Ratio(4, 2), pitches=leaf.pitches, **kw)]


def longfermata(leaf: Leaf, scale: Any) -> List[Leaf]:
    kw = _base_kwargs(leaf)
    return [Leaf(duration=leaf.duration * Ratio(6, 2), pitches=leaf.pitches, **kw)]


def verylongfermata(leaf: Leaf, scale: Any) -> List[Leaf]:
    kw = _base_kwargs(leaf)
    return [Leaf(duration=leaf.duration * Ratio(8, 2), pitches=leaf.pitches, **kw)]


# ============================================================
# Dispatch table
# ============================================================

ornament_function_map: Dict[str, Callable[[Leaf, Any], List[Leaf]]] = {
    "prall":           prall,
    "prallup":         prallup,
    "pralldown":       pralldown,
    "upprall":         upprall,
    "downprall":       downprall,
    "prallprall":      prallprall,
    "lineprall":       lineprall,
    "prallmordent":    prallmordent,
    "mordent":         mordent,
    "upmordent":       upmordent,
    "downmordent":     downmordent,
    "trill":           trill,
    "turn":            turn,
    "reverseturn":     reverseturn,
    "shortfermata":    shortfermata,
    "fermata":         fermata,
    "longfermata":     longfermata,
    "verylongfermata": verylongfermata,
}
